### PR TITLE
Prevent unwanted helper api calls when loading subscription notes

### PR DESCRIPTION
**Pull request by thilinah**
### All Submissions:

-   [ ] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

### Changes proposed in this Pull Request:

Connected WooCommerce stores send requests to helper API for fetching details related to subscriptions and to get information about plugin updates. The responses from these API endpoints are cached to prevent stores from sending multiple requests frequently. 

For example, the `GET subscriptions` request is cached for  1 hour and `POST update-check` request is cached for 12 hours.

But for some stores, the cache can be completely broken due to DB having a full disk. This is because when `set_transient` is called if the object cache is not configured for the site wp options table will be used to store the transient value. (further discussions in this [P2 post](https://wp.me/p6JqRr-4N1#comment-12553) and the [issue](https://github.com/Automattic/woocommerce.com/issues/11649)).

#### Problem

- Subscription notes are loaded via `admin_init`
- In order to prevent subscription notes from on every call to `admin_init` the option `LAST_REFRESH_OPTION_KEY` is used.
- But when the db disk is full, saving the latest updated time in wp_option should fail.
- This leads to subscription notes being loaded on every admin page load and cron run.
- Since the transients are not saved when the disk is full, this leads to store making get subscription API call multiple times.

Fix
- There is no need to check if subscription notes need to be refreshed on every admin page load. Instead, it can be checked only when the user visits a WC page showing the `Activity` tab.

Closes # .

### How to test the changes in this Pull Request:

- [ ] Have you followed the [Writing high-quality testing instructions guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions)?

#### Cron should not load subscription notes
- Make sure the site is connected to WCCOM via <your test site root>wp-admin/admin.php?page=wc-addons&section=helper
- Every time subscription notes are updated the option `woocommerce_admin-wc-helper-last-refresh` will be set to the current time. We will use this fact for testing.
- Checkout trunk.
- Run `wp option delete woocommerce_admin-wc-helper-last-refresh`
- Run cron `wp cron event run --all`
- Run `wp option get woocommerce_admin-wc-helper-last-refresh` and notice it exists.
- Checkout this branch.
- Run `wp option delete woocommerce_admin-wc-helper-last-refresh`
- Run cron `wp cron event run --all`
- Run `wp option get woocommerce_admin-wc-helper-last-refresh` and notice it doesn't exist.

#### Only visiting a WC Admin or a page with WC Admin Header Shall load subscription notes
- Checkout this branch.
- Run `wp option delete woocommerce_admin-wc-helper-last-refresh`
- Visit a non WC page like `wp-admin/edit-comments.php`
- Run `wp option get woocommerce_admin-wc-helper-last-refresh` and notice it doesn't exist.
- Visit a WC related page that shows `Activity` tab, such as `wp-admin/edit.php?post_type=shop_order`
Screenshot 2023-06-02 at 12 47 59
- Run `wp option get woocommerce_admin-wc-helper-last-refresh` and notice that it exists. (this means subscription notes are loading on this page)
- Repeat this test on any page to make sure all pages showing `Activity` tab are loading subscription notes.

#### Make sure the subscription notes are loading after this update
- Checkout this branch.
- Make sure your store is connected.
- You need to have an active product that is expiring soon (less than in 60 days)
- If this is not the case, you may update the expiry date of the subscription via the dashboard.
- Under [My Subscriptions](https://wp.test:4443/wp-admin/admin.php?page=wc-addons&section=helper) you should be able to see the expiry date.
Screenshot 2023-06-02 at 12 41 41
- Disable the auto-renew for this product via [My Subscriptions](https://woocommerce.test/my-account/my-subscriptions/)
- Run the following in `wp shell` to remove subscription notes.
```
\Automattic\WooCommerce\Admin\Notes\Notes::delete_notes_with_name('wc-admin-wc-helper-subscription');
```
- Update subscriptions under [My Subscriptions](https://wp.test:4443/wp-admin/admin.php?page=wc-addons&section=helper) 
Screenshot 2023-06-02 at 12 41 55
- Run `wp option delete woocommerce_admin-wc-helper-last-refresh` reset last subscription notes refresh time.
- Load any WC page.
- Notice there is a new subscription note as a reminder to enable auto-renew for the above mentioned product.
Screenshot 2023-06-02 at 12 41 21

### Other information:

-   [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [ ] Have you created a changelog file for each project being changed, ie `pnpm --filter=<project> changelog add`?
-   [ ] Have you included testing instructions?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: plugins/woocommerce/src/Internal/Admin/Notes/WooSubscriptionsNotes.php
Comment (vedanshujain) [inline plugins/woocommerce/src/Internal/Admin/Notes/WooSubscriptionsNotes.php]: seems like this isn't required either, perhaps we can move this below the early return statements?

**Review by vedanshujain — DISMISSED** [review]
Requesting a minor change if its possible

**Comment by vedanshujain** [comment]
@thilinah This looks good to me, but I am assuming you are waiting for wc-admin folks to also take a look?

**Review by rcstr — DISMISSED** [review]
Changes looks good and work as expected

**Comment by thilinah** [comment]
> @thilinah This looks good to me, but I am assuming you are waiting for wc-admin folks to also take a look?

@vedanshujain if the changes look good can you please approve the PR, I see it's blocked due to the change request.

**Comment by vedanshujain** [comment]
@thilinah dismissed my review, should be mergeable now

**Review by nirolph — DISMISSED** [review]

**Comment by denho** [comment]
@thilinah The WC notes are shown not only on `wc-admin` and `wc-addons` pages. They are also loaded on other WC pages in admin, such as orders list or settings.  
image
On those pages, the subscription notes are not refreshed due to changes in your PR, and the user will not get the latest  notes until they visit `wc-admin` and `wc-addons` pages.

**Comment by thilinah** [comment]
> On those pages, the subscription notes are not refreshed due to changes in your PR, and the user will not get the latest notes until they visit wc-admin and wc-addons pages.

@denho in this case I think skipping subscription notes loading based on the page won't work. I've noticed that other WC plugins like subscriptions and bookings also show the `Activity` button. So post_types such as `shop_order`, `shop_subscription`, `shop_coupon`, `product`. 

Due to this, I think we should consider moving forward with https://github.com/woocommerce/woocommerce/pull/36846 and move the discussion to that PR for refining the approach.

**Comment by thilinah** [comment]
Also, another way could be to make sure subscription notes are not loaded via the cron.

**Comment by thilinah** [comment]
@denho @nirolph @rcstr re-requesting reviews for this PR since the code and test instructions are updated.

**Review by rcstr — APPROVED** [review]
Changes looks good and works as described

**Comment by jonathansadowski** [comment]
I'm going to clear the milestone from this for now, so that this doesn't inadvertently get missed once it's merged. A milestone will be added automatically by our automation upon merge.

File: plugins/woocommerce/src/Internal/Admin/Notes/WooSubscriptionsNotes.php
Comment (denho) [inline plugins/woocommerce/src/Internal/Admin/Notes/WooSubscriptionsNotes.php]: How about adding `use \Automattic\WooCommerce\Admin\PageController` and replacing this line with 
```suggestion
		if ( ! PageController::is_admin_or_embed_page() ) {
```

File: plugins/woocommerce/src/Internal/Admin/Notes/WooSubscriptionsNotes.php
Comment (denho) [inline plugins/woocommerce/src/Internal/Admin/Notes/WooSubscriptionsNotes.php]: Would be good to have a more meaningful description of the method functionality in the DocBlock, and method name.

**Review by denho — APPROVED** [review]
Added some suggestions in the code. Otherwise, LGTM!

**Comment by thilinah** [comment]
@jonathansadowski, the PR is approved. Is it possible to get this merged?

**Comment by jonathansadowski** [comment]
Hi @thilinah, please feel free to merge once checks have passed (as they have in this case), and you've gotten your PR approved. If you don't have access to merge yourself, let me know and I can take a look further.

**Comment by thilinah** [comment]
@jonathansadowski I'm not authorized to merge.

**Comment by jonathansadowski** [comment]
Thanks @thilinah, I've merged this for now — I'll look into whether or not that should be the case.

**Pull request merged**
